core/arch/arm/include/sm/optee_smc.h

```diff
@@ -93,6 +93,12 @@
 #define OPTEE_SMC_SHM_CACHED		1
 
 /*
+ * Privileged call flag. Used by supplicant. Asks OP-TEE to use reserved
+ * thread if all regular threads are used.
+ */
+#define OPTEE_SMC_FLAGS_PRIVILEGED	BIT(0)
```

Comment:
Ah, I see. We need 3 pages per thread for stack, right? This feature will require one extra thread, so additional 12Kb of memory will be used. Is this too much?

I can imagine another approach: save context (and stack) of a suspended thread to dynamically allocated buffer and use that thread for a privileged call. But what if we couldn't allocate buffer for context? There will be deadlock.

Another approach is to try to squeeze privileged thread stack into one of the suspended thread's stack. But this is also prone to problems with the stack size.

Probably we can forbid STD SMC calls from a RPC call. Then there will be no need to a privileged thread. But in that case we will need another call path to register SHM in a RPC request. I had that solution (maybe you remember `OPTEE_MSG_ATTR_TYPE_NEXT_NESTED`) in earlier versions of the pacthes. But there was the extra param type (`NESTED` one), additional logic, etc, etc. I didn't liked it, so I choose simpler approach with existing shared memory calls.

Maybe you have ideas on how to avoid deadlock?